### PR TITLE
Add Extending Tuner Chapter

### PR DESCRIPTION
**Pull request by be-marc**

### CODE REVIEW

File: bookdown/06-extending-tuner.Rmd [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+In this section, we show how to implement a custom tuner for `r mlr_pkg("mlr3tuning")`.
+
+### Adding a new Tuner via bbotk
+
+The `r mlr_pkg("bbotk")` package provides black box optimization algorithms.
+If the new custom tuner is not restricted to hyperparameter tuning, it should be implemented in `r mlr_pkg("bbotk")`.
+The `bbotk::Optimizer` can be easily transformed to a `mlr3tuning::Tuner` later.
+
+1. Check the tuner does not already exist as a `bbotk::Optimizer` or `mlr3tuning::Tuner`.
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:11]: Maybe a small point but is it obvious to a user how they can check this? Like is there specific listing functionality or optimizers or should they check GH?

File: bookdown/06-extending-tuner.Rmd [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+2. Use one of the existing optimizer as a [template](#tuner-template).
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:12]: "existing optimizers"

File: bookdown/06-extending-tuner.Rmd [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+3. Add [`.optimize`](#tuner-optimize) private method to optimizer.
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:13]: Should this not be overwrite `.optimize` if using an existing one as a template?

File: bookdown/06-extending-tuner.Rmd [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+4. Optionally, overwrite the default [`.assign_result`](#tuner-add-result) private method.
+5. Use the [`mlr3tuning::TunerFromOptimizer`](#tuner-from-optimizer) class to transform the `bbotk::Optimizer` to a `mlr3tuning::Tuner`.
+6. Add [unit tests](#tuner-test) for the optimizer and the tuner.
+7. Open a new pull request for the [`bbotk::Optimizer`](https://github.com/mlr-org/bbotk/pulls) and a second one for the [`mlr3tuning::Tuner`](https://github.com/mlr-org/mlr3tuning/pulls).
+
+### Adding a new Tuner via mlr3tuning
+
+If the new custom tuner is only usable for hyperparameter tuning, it should be directly implemented in `r mlr_pkg("mlr3tuning")`.
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:21]: What's the alternative to only hyperparameter tuning?

File: bookdown/06-extending-tuner.Rmd [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+1. Check the tuner does not already exist as a `bbotk::Optimizer` or `mlr3tuning::Tuner`.
+2. Use one of the existing tuners as a [template](#tuner-template).
+3. Add [`.optimize`](#tuner-optimize) private method to tuner.
+4. Optionally, overwrite the default [`.assign_result`](#tuner-add-result) private method.
+5. Add [unit tests](#tuner-test).
+6. Open a new pull request for the new [tuner](https://github.com/mlr-org/mlr3tuning/pulls).
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:28]: duplicate of above

File: bookdown/06-extending-tuner.Rmd [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+### Template {#tuner-template}
+
+If the new custom tuner is implemented via `r mlr_pkg("bbotk")`, use one of the existing optimizer as a template e.g. [`bbotk::OptimizerRandomSearch`](https://github.com/mlr-org/bbotk/blob/master/R/OptimizerRandomSearch.R). There are currently only two tuners that are not based on a `bbotk::Optimizer`: [`mlr3hyperband::TunerHyperband`](https://github.com/mlr-org/mlr3hyperband/blob/master/R/TunerHyperband.R) and [`mlr3tuning::TunerIrace`](https://github.com/mlr-org/mlr3tuning/blob/master/R/TunerIrace.R). Both are rather complex but you can still use the documentation and class structure as a template. The following steps are identical for optimizers and tuners.
+
+Rewrite the meta information in the documentation and create a new class name.
+Scientific sources can be added in `R/bibentries.R` which are added under `@source` in the documentation.
+The example and parts of the documentation are auto-generated based on the `@templateVar id <tuner_id>`.
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:36]: I think this is confusing unless you specify which 'parts'

File: bookdown/06-extending-tuner.Rmd [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+The `$.optimize()` private method takes an instance, proposes new points and calls the `$eval_batch()` method of the instance to evaluate them.
+Usually, this is the main part of the custom tuner.
+Optimization functions from external packages usually take an objective function as an argument. 
+In this case, you can pass `inst$objective_function` which internally calls `$eval_batch()`.
+Check out [`OptimizerGenSA`](https://github.com/mlr-org/bbotk/blob/master/R/OptimizerGenSA.R) for an example.
+
+### Assign result method {#tuner-add-result}
+
+The default `$.assign_result()` private method simply obtains the best performing result from the archive.
+The default method can be overwritten if the new tuner determines the result of the optimization in a different way.
```
Comment (RaphaelS1) [inline bookdown/06-extending-tuner.Rmd:51]: Might be nice to add an example of a tuner that does this

File: bookdown/02-basics-tasks.Rmd [hunk 8]
```diff
@@ -80,6 +80,7 @@ mlr_tasks
 We can get a more informative summary of the example tasks by converting the dictionary to a `r ref("data.table()")` object:
 
 ```{r 02-basics-tasks-005}
+library(data.table)
```
Comment (jakob-r) [inline bookdown/02-basics-tasks.Rmd:83]: why is this added here?

File: bookdown/06-extending-tuner.Rmd (hunk 1 above)
Comment (be-marc) [reply bookdown/06-extending-tuner.Rmd]: Added links to gh repositories.

File: bookdown/06-extending-tuner.Rmd (hunk 4 above)
Comment (be-marc) [reply bookdown/06-extending-tuner.Rmd]: I rewrote the introduction. Maybe it is clearer now?

File: bookdown/06-extending-tuner.Rmd (hunk 5 above)
Comment (be-marc) [reply bookdown/06-extending-tuner.Rmd]: Simplified this to one list.

File: bookdown/02-basics-tasks.Rmd (hunk 8 above)
Comment (be-marc) [reply bookdown/02-basics-tasks.Rmd:83]: The book does not render without this on my machine. But if it just me, we can remove it.

**Comment by mllg** [comment]
Ready for merge?

**Review by jakob-r — APPROVED** [review]
